text/0000-safer-transmute.md

```diff
@@ -0,0 +1,1796 @@
+        {
+            type Archetype = &'a T::Archetype;
+        }
+
+        impl<'a, T: ?Sized> PromiseTransmutableFrom for &'a T
+        where
+            T: PromiseTransmutableFrom,
+            &'a T::Archetype
+                : TransmuteInto<&'a T, NeglectStability>
+                + PromiseTransmutableFrom,
+        {
+            type Archetype = &'a T::Archetype;
+        }
+
+        impl<'a, T: ?Sized> PromiseTransmutableInto for &'a mut T
+        where
+            T: PromiseTransmutableInto,
+            &'a mut T::Archetype
+                : TransmuteFrom<&'a mut T, NeglectStability>
+                + PromiseTransmutableInto,
+        {
+            type Archetype = &'a mut T::Archetype;
+        }
+
+        impl<'a, T: ?Sized> PromiseTransmutableFrom for &'a mut T
+        where
+            T: PromiseTransmutableFrom,
+            &'a mut T::Archetype
+                : TransmuteInto<&'a mut T, NeglectStability>
+                + PromiseTransmutableFrom,
+        {
+            type Archetype = &'a mut T::Archetype;
+        }
+    }
+
+    /// Static checks that may be neglected when determining if a type is `TransmuteFrom` some other type.
+    pub mod options {
+
+        /// Options that may be used with safe transmutations.
+        pub trait SafeTransmuteOptions: TransmuteOptions {}
+
+        /// Options that may be used with unsafe transmutations.
+        pub trait TransmuteOptions: private::Sealed {}
+
+        impl SafeTransmuteOptions for () {}
+        impl TransmuteOptions for () {}
+
+        /// Neglect the stability check of `TransmuteFrom`.
+        pub struct NeglectStability;
+
+        // Uncomment this if/when constructibility is fully implemented:
+        // impl SafeTransmuteOptions for NeglectStability {}
+        impl TransmuteOptions for NeglectStability {}
+
+        // prevent third-party implementations of `TransmuteOptions`
+        mod private {
+            use super::*;
+
+            pub trait Sealed {}
+
+            impl Sealed for () {}
+            impl Sealed for NeglectStability {}
+        }
+    }
+}
+```
+
+### Towards an Initial, Smart Implementation
+
+To support transmutations between different types, implementers of this RFC should begin by defining a `transmute_from` lang item to annotate libcore's definition of `TransmuteFrom`. Whether `TransmuteFrom` is implemented for a given type and parameters shall be determined within the implementation of the type system (à la [`Sized`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyS.html#method.is_sized) and [`Freeze`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyS.html#method.is_freeze)).
+
+This initial smart implementation may be made simpler by:
+  - *not* supporting enums
+  - *not* supporting unions
+  - *not* supporting `Neglect` parameters besides `()` and `NeglectStability`
+  - simplifying constructability
+
+#### Simplifying Constructability
```

Comment:
### Update

The hand-waving about how the fully implicit constructability check might be implemented in the trait system isn't very satisfying, nor is the soundness hole created by the simplified formulation of the check. We've been considering an alternative solution [on zulip](https://rust-lang.zulipchat.com/#narrow/stream/216762-project-safe-transmute/topic/Stability.20and.20visibility/near/209566004) that doesn't sacrifice soundness; to summarize:

Introduce:
- a macro, `Here!()`, that produces a type which uniquely identifies its invocation scope, and
- a trait, `Constructible<Scope>`, which is implemented for a type `T` if `T` is fully implicitly constructible in the scope represented by `Scope`.

To illustrate how this would all fit together, imagine a built-in trait, `WellDefinedTransmuteFrom<Src>` that implements *just* the rules in *When is a transmutation well-defined?*. Given this:

A trait that implements the rules described in *When is a transmutation safe?* would be defined as:
```rust
// is the transmutation well-defined and safe?
pub unsafe trait SafeTransmuteFrom<Src, Scope> {}

unsafe impl<Src, Dst, Scope> SafeTransmuteFrom<Src, Scope> for Dst
where
    Src: Constructible<Scope>,
    Dst: Constructible<Scope> + SoundTransmuteFrom<Src>,
{}
```

A trait that implements the rules described in *When is a transmutation stable?* would be defined as:
```rust
// Is the transmutation well-defined and stable?
pub unsafe trait StableTransmuteFrom<Src, Scope> {}

unsafe impl<Src, Dst, Scope> StableTransmuteFrom<Src, Scope> for Dst
where
    Src: PromiseTransmutableInto<Scope>,
    Dst: PromiseTransmutableFrom<Scope>,
    Dst::Archetype: SafeTransmuteFrom<Src::Archetype, Scope>,
{}
```

For users building highly-generic abstractions over `TransmuteFrom`, this requires adds a chore of threading of an extra `Scope` parameter through their traits.

However, most actual transmutations could basically just ignore this parameter, especially with the help of a macro:
```rust
pub macro transmute($expr: expr) {
    TransmuteFrom::<_, Here!()>::transmute_from($expr)
}
```

I *think* this is more feasible to implement than the RFC's current speculative "make the trait system aware of scope somehow" gambit, but I'd love feedback from the compiler team. 

It'll take me a few days to write this up, but I expect will considerably shorten the *Reference Level Explanation*.